cpp/include/cudf/column/column_device_view.cuh

```diff
@@ -500,6 +531,124 @@ class alignas(16) column_device_view : public detail::column_device_view_base {
   using const_pair_rep_iterator =
     thrust::transform_iterator<detail::pair_rep_accessor<T, has_nulls>, count_it>;
 
+  /**
+   * @brief Return an optional iterator to the first element of the column.
+   *
+   * Dereferencing the returned iterator returns a `thrust::optional<T>`.
+   *
+   * When the element of an iterator contextually converted to bool, the conversion returns true
+   * if the object contains a value and false if it does not contain a value.
+   *
+   * optional_begin with mode `DYNAMIC` defers the assumption of nullability to
+   * runtime, with the user stating on construction of the iterator if column has nulls.
+   * `DYNAMIC` mode is nice when an algorithm is going to execute on mutliple
+   * iterators and you don't want to compile all the combinations of iterator types
+   *
+   * Example:
+   *
+   * \code{.cpp}
+   * template<typename T>
+   * void some_function(cudf::column_view<T> const& col_view, bool has_nulls){
+   *    auto d_col = cudf::column_device_view::create(col_view);
+   *    // Create a `DYNAMIC` optional iterator
+   *    auto optional_iterator = d_col->optional_begin<T>(cudf::contains_nulls::DYNAMIC{},
+   *                                                      has_nulls);
+   * }
```

Comment:
```suggestion
   * template<typename T>
   * void some_function(cudf::column_view<T> const& col_view){
   *    auto d_col = cudf::column_device_view::create(col_view);
   *    // Create a `DYNAMIC` optional iterator
   *    auto optional_iterator = d_col->optional_begin<T>(cudf::contains_nulls::DYNAMIC{},
   *                                                      col_view.has_nulls());
   * }
```
I feel this may better communicate the expectation of the `has_nulls` parameter since the code does check the parameter is consistent with the `d_col->nullable()` member.